reference-implementation/lib/readable-byte-stream.js

```diff
@@ -0,0 +1,965 @@
+      throw new TypeError(
+          'ReadableByteStreamController instances can only be created by the ReadableByteStream constructor');
+    }
+
+    this._controlledReadableByteStream = controlledReadableByteStream;
+
+    this._underlyingByteSource = underlyingByteSource;
+
+    this._considerReissueUnderlyingByteSourcePullOrPullInto = false;
+    this._insideUnderlyingByteSource = false;
+
+    this._pendingPullIntos = [];
+
+    this._queue = [];
+    this._totalQueuedBytes = 0;
+
+    InvokeOrNoop(underlyingByteSource, 'start', [this]);
+  }
+
+  close() {
+    if (!IsReadableByteStreamController(this)) {
+      throw new TypeError(
+          'ReadableByteStreamController.prototype.close can only be used on a ReadableByteStreamController');
+    }
+
+    const stream = this._controlledReadableByteStream;
+
+    if (stream._closeRequested) {
+      throw new TypeError('The stream has already been closed; do not close it again!');
+    }
+    if (stream._state !== 'readable') {
+      throw new TypeError('The stream is not in the readable state and cannot be closed');
+    }
+
+    if (this._totalQueuedBytes > 0) {
+      this._closeRequested = true;
+
+      return;
+    }
+
+    const reader = stream._reader;
+
+    if (reader === undefined || IsReadableByteStreamReader(reader)) {
+      CloseReadableByteStream(stream);
+      return;
+    }
+
+    assert(IsReadableByteStreamByobReader(reader), 'reader must be ReadableByteStreamByobReader');
+
+    if (this._pendingPullIntos.length > 0 && this._pendingPullIntos[0].bytesFilled > 0) {
+      DestroyReadableByteStreamController(this);
+      const e = new TypeError('Insufficient bytes to fill elements in the given buffer');
+      ErrorReadableByteStream(stream, e);
+
+      throw e;
+    }
+
+    CloseReadableByteStream(stream);
+  }
+
+  enqueue(chunk) {
+    if (!IsReadableByteStreamController(this)) {
+      throw new TypeError(
+          'ReadableByteStreamController.prototype.enqueue can only be used on a ReadableByteStreamController');
+    }
+
+    const stream = this._controlledReadableByteStream;
+
+    if (stream._closeRequested) {
+      throw new TypeError('stream is closed or draining');
+    }
+    if (stream._state !== 'readable') {
+      throw new TypeError('The stream is not in the readable state and cannot be enqueued to');
+    }
+
+    const reader = stream._reader;
+
+    if (reader === undefined) {
+      EnqueueInReadableByteStreamController(this, chunk);
+      return;
+    }
+
+    if (IsReadableByteStreamReader(reader)) {
+      if (reader._readRequests.length === 0) {
+        EnqueueInReadableByteStreamController(this, chunk);
+        return;
+      }
+
+      assert(this._queue.length === 0);
+
+      const req = reader._readRequests.shift();
+      // TODO: Detach chunk.
+      req.resolve(CreateIterResultObject(chunk, false));
+
+      if (this._closeRequested === true) {
+        CloseReadableByteStream(stream);
```

Comment:
This is really confusing for me because of the _closeRequested issue. It seems now like there are two different _closeRequested members---one for the stream, and one for the controller? They should probably be renamed...